src/libappimage/libappimage.cpp

```diff
@@ -0,0 +1,364 @@
+
+        return result;
+    } catch (const std::runtime_error& err) {
+        std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+    } catch (...) {
+        std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+    }
+
+    // Create empty string list
+    result = static_cast<char**>(malloc(sizeof(char*)));
+    result[0] = nullptr;
+
+    return result;
+}
+
+void appimage_string_list_free(char** list) {
+    for (char** ptr = list; ptr != NULL && *ptr != NULL; ptr++)
+        free(*ptr);
+
+    free(list);
+}
+
+bool
+appimage_read_file_into_buffer_following_symlinks(const char* appimage_file_path, const char* file_path, char** buffer,
+                                                  unsigned long* buf_size) {
+    // Init output params
+    *buffer = nullptr;
+    *buf_size = 0;
+
+    std::string targetFile = file_path;
+    std::vector<std::string> visitedEntries;
+
+    while (!targetFile.empty()) {
+        // Find loops
+        if (std::find(visitedEntries.begin(), visitedEntries.end(), targetFile) != visitedEntries.end()) {
+            std::clog << "WARNING: Links loop found while extracting " << file_path;
+            break;
+        }
+
+        visitedEntries.emplace_back(targetFile);
+        std::string nextHoop;
+        try {
+            AppImage appImage(appimage_file_path);
+
+            for (auto itr = appImage.files(); itr != itr.end(); ++itr) {
+                if (itr.path() == targetFile) {
+                    if (itr.type() == PayloadEntryType::REGULAR) {
+                        auto data = std::vector<char>(std::istreambuf_iterator<char>(itr.read()),
+                                                      std::istreambuf_iterator<char>());
+
+                        *buffer = static_cast<char*>(malloc(sizeof(char) * data.size()));
+                        memcpy(*buffer, data.data(), data.size());
+
+                        *buf_size = data.size();
+                        return true;
+
+                    } else nextHoop = itr.linkTarget();
+
+                    break;
+                }
+            }
+
+
+            if (!nextHoop.empty())
+                targetFile = nextHoop;
+        } catch (const std::runtime_error& err) {
+            std::clog << "Error at " << __FUNCTION__ << " : " << err.what() << std::endl;
+        } catch (...) {
+            std::clog << "Error at " << __FUNCTION__ << " : that's all we know." << std::endl;
+        }
+    }
+    return false;
+}
+
+void appimage_extract_file_following_symlinks(const char* appimage_file_path, const char* file_path,
+                                              const char* target_file_path) {
+
+    std::string targetFile = file_path;
+    std::vector<std::string> visitedEntries;
+
+    while (!targetFile.empty()) {
+        // Find loops
+        if (std::find(visitedEntries.begin(), visitedEntries.end(), targetFile) != visitedEntries.end()) {
```

Comment:
See the comment for the other "loop detection". All this loop code is implemented here more than once. Redundancy warning, please refactor and remove the redundancy.

Reply:
Should be fixed at #66